src/orange/MatrixUtils.cc

```diff
@@ -0,0 +1,197 @@
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Normalize and orthogonalize a small, dense matrix.
+ *
+ * \return normalization (zero if already orthonormal)
+ *
+ * This is used for constructing rotation matrices from user-given matrices
+ * that may only have a few digits of precision (e.g. were read from an XML
+ * file). It uses the modified Gram-Schmidt orthogonalization algorithm.
+ */
+template<class T, size_type N>
+void orthonormalize(SquareMatrix<T, N>* mat)
+{
+    for (size_type i = 0; i != N; ++i)
+    {
+        Array<T, N>& cur = (*mat)[i];
+
+        // Orthogonalize against previous rows
+        for (size_type ip = 0; ip != i; ++ip)
+        {
+            Array<T, N>& prev = (*mat)[ip];
+            T proj = dot_product(cur, prev);
+            axpy(-proj, prev, &cur);
+        }
+
+        // Normalize row
+        T inv_mag = 1 / norm(cur);
+        for (size_type j = 0; j != N; ++j)
+        {
+            cur[j] *= inv_mag;
+        }
+    }
+
+    // Check result for orthonormality: if the matrix is singular this will
+    // fail
+    if (CELERITAS_DEBUG)
+    {
+        for (size_type i = 0; i != N; ++i)
+        {
+            for (size_type ip = 0; ip != i; ++ip)
+            {
+                CELER_ENSURE(soft_equal(dot_product((*mat)[i], (*mat)[ip]),
+                                        static_cast<T>(i == ip ? 1 : 0)));
+            }
+        }
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create a C-ordered rotation matrix.
+ */
+Mat3 make_rotation(Axis ax, Turn theta)
+{
+    CELER_EXPECT(ax < Axis::size_);
+
+    // Calculate sin and cosine with less precision loss
+    real_type cost;
+    real_type sint;
+    sincos(theta, &sint, &cost);
+
+    // Avoid signed zeros with this implementation
+    real_type msint = real_type(0) - sint;
+
+    // Fill result with zeros
+    Mat3 r;
+    r.fill(Real3{0, 0, 0});
+
+    // {i, i} gets 1
+    r[to_int(ax)][to_int(ax)] = 1;
+
+    switch (ax)
+    {
+        case Axis::x:
+            r[1][1] = cost;
+            r[1][2] = msint;
+            r[2][1] = sint;
+            r[2][2] = cost;
+            break;
+        case Axis::y:
+            r[0][0] = cost;
+            r[0][2] = sint;
+            r[2][0] = msint;
+            r[2][2] = cost;
+            break;
+        case Axis::z:
+            r[0][0] = cost;
+            r[0][1] = msint;
+            r[1][0] = sint;
+            r[1][1] = cost;
+            break;
+        default:
+            CELER_ASSERT_UNREACHABLE();
+    }
+    return r;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Rotate a C-ordered rotation matrix.
+ *
+ * This applies the new axis + turn as a rotation operator to the left of
+ * the matrix.
+ *
+ * Rotate first by1/3
+ * \example
+ * auto r = make_rotation(Axes::x, Turn{0.3}, make_rotation(Axes::y, -0.6));
+ * \example
```

Comment:
```suggestion
 * Rotate first by1/3
 * \example
 * auto r = make_rotation(Axis::x, Turn{0.3}, make_rotation(Axis::y, -0.6));
 * \example
```
Plus `\code` and `\endcode` minus 🐿️